src/sinks/appsignal/mod.rs

```diff
@@ -0,0 +1,265 @@
+struct AppsignalDefaultBatchSettings;
+
+impl SinkBatchSettings for AppsignalDefaultBatchSettings {
+    const MAX_EVENTS: Option<usize> = Some(100);
+    const MAX_BYTES: Option<usize> = Some(450_000);
+    const TIMEOUT_SECS: f64 = 1.0;
+}
+
+impl_generate_config_from_default!(AppsignalSinkConfig);
+
+#[async_trait::async_trait]
+impl SinkConfig for AppsignalSinkConfig {
+    async fn build(
+        &self,
+        cx: SinkContext,
+    ) -> crate::Result<(super::VectorSink, super::Healthcheck)> {
+        let push_api_key = self.push_api_key.inner().to_string();
+        let request_settings = self.request.unwrap_with(&TowerRequestConfig::default());
+        let batch_settings = self.batch.into_batch_settings()?;
+
+        let buffer = JsonArrayBuffer::new(batch_settings.size);
+
+        let tls_settings = TlsSettings::from_options(&None)?;
+        let client = HttpClient::new(tls_settings, cx.proxy())?;
+
+        let sink = BatchedHttpSink::new(
+            self.clone(),
+            buffer,
+            request_settings,
+            batch_settings.timeout,
+            client.clone(),
+        )
+        .sink_map_err(|error| error!(message = "Fatal AppSignal sink error.", %error));
+
+        let healthcheck = healthcheck(
+            endpoint_uri(self.endpoint.clone(), "vector/healthcheck")?,
+            push_api_key,
+            client,
+        )
+        .boxed();
+
+        Ok((super::VectorSink::from_event_sink(sink), healthcheck))
+    }
+
+    fn input(&self) -> Input {
+        Input::new(DataType::Metric | DataType::Log)
+    }
+
+    fn acknowledgements(&self) -> &AcknowledgementsConfig {
+        &self.acknowledgements
+    }
+}
+
+pub struct AppsignalEventEncoder {
+    transformer: Transformer,
+}
+
+impl HttpEventEncoder<serde_json::Value> for AppsignalEventEncoder {
+    fn encode_event(&mut self, mut event: Event) -> Option<serde_json::Value> {
+        self.transformer.transform(&mut event);
+
+        match event {
+            Event::Log(log) => Some(json!({ "log": log })),
+            Event::Metric(metric) => Some(json!({ "metric": metric })),
+            _ => panic!("The AppSignal sink does not support this event: {event:?}"),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl HttpSink for AppsignalSinkConfig {
+    type Input = serde_json::Value;
+    type Output = Vec<BoxedRawValue>;
+    type Encoder = AppsignalEventEncoder;
+
+    fn build_encoder(&self) -> Self::Encoder {
+        AppsignalEventEncoder {
+            transformer: self.encoding.clone(),
+        }
+    }
+
+    async fn build_request(&self, events: Self::Output) -> crate::Result<Request<Bytes>> {
+        let uri = endpoint_uri(self.endpoint.clone(), "vector/events")?;
+        let request = Request::post(uri)
+            .header(
+                AUTHORIZATION,
+                format!("Bearer {}", self.push_api_key.inner()),
+            )
+            .header("Content-Encoding", "gzip");
```

Comment:
Would it be reasonable to allow configuration of the compression through our struct `Compression` , that has gzip as one of the options, and default to that in the configuration struct?

Reply:
Ah, this type of thing is a known issue on a cross cutting scale that hopefully we can resolve in the future. But as it stands, the way to do this is just define a specific `Compression` enum for this sink. For example, the `aws_kinesis_firehose` does this.

Reply:
The aws_kinesis_firehose sink uses the aws [base config struct](https://github.com/vectordotdev/vector/blob/8afce37608b95b18dd97be6aca8fd6a46d026eaa/src/sinks/aws_kinesis/config.rs#L48) that uses the compression util struct. It doesn't have a custom one.
The [kafka sink uses this KafkaCompression struct](https://github.com/vectordotdev/vector/blob/8afce37608b95b18dd97be6aca8fd6a46d026eaa/src/kafka.rs#L20-L40), but without compression levels like the util compresssion struct has.

I asked my team, and for us it's not needed for users of our sink to specify the specific compression type. I don't know how often it's expected people change the compression for an external service sink like ours.

If it is required, I suggest going for the approach in the diff below (so I don't need to write custom serializers and deserializers like for the Compression struct) along with how the [http sink compresses the payload](https://github.com/vectordotdev/vector/blob/8afce37608b95b18dd97be6aca8fd6a46d026eaa/src/sinks/http.rs#L378-L396) (which where I got the current gzip compression from) using the default compression levels.

<details>
  <summary>Click to expand Git diff</summary>

  ```diff
  diff --git src/sinks/appsignal/mod.rs src/sinks/appsignal/mod.rs
  index 3db6afc5f..c5912f1f9 100644
  --- src/sinks/appsignal/mod.rs
  +++ src/sinks/appsignal/mod.rs
  @@ -55,6 +55,24 @@ pub const fn as_error_type(&self) -> &'static str {
      }
  }

  +/// Supported compression types for the AppSignal sink.
  +#[configurable_component]
  +#[derive(Copy, Clone, Debug, Derivative, Eq, PartialEq)]
  +#[serde(rename_all = "lowercase")]
  +#[derivative(Default)]
  +enum AppsignalCompression {
  +    /// [Gzip][gzip] compression.
  +    ///
  +    /// [gzip]: https://www.gzip.org/
  +    #[derivative(Default)]
  +    Gzip,
  +
  +    /// [Zlib][zlib] compression.
  +    ///
  +    /// [zlib]: https://zlib.net/
  +    Zlib,
  +}
  +
  /// Configuration for the `appsignal` sink.
  #[configurable_component(sink("appsignal"))]
  #[derive(Clone, Debug, Default)]
  @@ -70,6 +88,10 @@ pub struct AppsignalSinkConfig {
      #[configurable(metadata(docs::examples = "${APPSIGNAL_PUSH_API_KEY}"))]
      push_api_key: SensitiveString,

  +    #[configurable(derived)]
  +    #[serde(default)]
  +    compression: AppsignalCompression,
  +
      #[configurable(derived)]
      #[serde(default)]
      batch: BatchConfig<AppsignalDefaultBatchSettings>,
  ```
</details>

Let me know if is required and what solution you prefer.

Reply:
Ah, my apologies- I was looking at the `aws_kinesis_firehose` source, as opposed to the sink, and didn't specify. And actually that example I was looking at is the compression for records as opposed to the whole request, but it was meant to serve as an example of defining an enum with a subset of the compression levels, which you already pointed to.
(https://github.com/vectordotdev/vector/blob/112358e69ed937d0751a4e70af1cc57a5f1cf4f0/src/sources/aws_kinesis_firehose/mod.rs#L99-L124)

What you have outlined in that diff is what I had in mind (and using the match logic with the builder as you pointed to the `http` sink doing). My inclination is that if different compression schemes are supported by the destination, and we are to default to using one, then the remaining should be configurable, as they are in the other sinks.

Reply:
I've updated the sink to support the zlib compression 👍